plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1221 @@
+		}
 
-func (p *PowerStat) initRaplService() {
-	if p.packageCurrentPowerConsumption || p.packageCurrentDramPowerConsumption || p.packageThermalDesignPower || p.packageTurboLimit ||
-		p.packageUncoreFrequency || p.packageCPUBaseFrequency {
-		p.rapl = newRaplServiceWithFs(p.Log, p.fs)
+		// One or more modules, needed to get metrics, failed to initialize. The plugin continues its execution, and it will not
+		// gather metrics relying on these modules. Instead, logs the error message including module names that failed to initialize.
+		p.Log.Warnf("Plugin started with errors: %v", err)
 	}
+
+	return nil
 }
 
-// fill CPUBusClockValue if required
-func (p *PowerStat) fillCPUBusClock() {
-	if p.packageCPUBaseFrequency {
-		// cpuBusClock is the same for every core/socket.
-		busClockInfo := p.getBusClock("0")
-		if busClockInfo == 0 {
-			p.Log.Warn("Disabling package metric: cpu_base_frequency_mhz. Can't detect bus clock value")
-			p.packageCPUBaseFrequency = false
-			return
-		}
+// Stop deactivates perf events if one or more of the requested metrics rely on perf.
+func (p *PowerStat) Stop() {
+	if !p.needsPerf {
+		return
+	}
 
-		p.cpuBusClockValue = busClockInfo
+	if err := p.fetcher.DeactivatePerfEvents(); err != nil {
+		p.Log.Errorf("Failed to deactivate perf events: %v", err)
 	}
 }
 
-// Gather takes in an accumulator and adds the metrics that the Input gathers
+// Gather collects the plugin's metrics.
 func (p *PowerStat) Gather(acc telegraf.Accumulator) error {
-	if p.areGlobalMetricsEnabled() {
-		p.addGlobalMetrics(acc)
+	// gather CPU metrics relying on coreFreq and msr which share CPU IDs.
+	if p.needsCoreFreq || p.needsMsrCPU {
+		p.addCPUMetrics(acc)
 	}
 
-	if p.areCoreMetricsEnabled() {
-		if p.msr.isMsrLoaded() {
-			p.logOnce["msr"] = nil
-			p.addPerCoreMetrics(acc)
-		} else {
-			err := errors.New("Error while trying to read MSR (probably msr module was not loaded)")
-			if val := p.logOnce["msr"]; val == nil || val.Error() != err.Error() {
-				p.Log.Errorf("%v", err)
-				// Remember that specific error occurs to omit logging next time
-				p.logOnce["msr"] = err
-			}
-		}
+	// gather CPU metrics relying on perf.
+	if p.needsPerf {
+		p.addCPUPerfMetrics(acc)
 	}
 
-	// Gathering the first iteration of metrics was skipped for most of them because they are based on delta calculations
-	p.skipFirstIteration = false
+	// gather package metrics.
+	if len(p.PackageMetrics) != 0 {
+		p.addPackageMetrics(acc)
+	}
 
 	return nil
 }
 
-func (p *PowerStat) addGlobalMetrics(acc telegraf.Accumulator) {
-	// Prepare RAPL data each gather because there is a possibility to disable rapl kernel module
-	p.rapl.initializeRaplData()
-	for socketID := range p.rapl.getRaplData() {
-		if p.packageTurboLimit {
-			p.addTurboRatioLimit(socketID, acc)
-		}
+// parseConfig is a helper method that parses configuration fields from the receiver such as included/excluded CPU IDs.
+func (p *PowerStat) parseConfig() error {
```

Comment:
I am aware `parseConfig` is only called once, but based on this [comment](https://github.com/influxdata/telegraf/pull/14363#discussion_r1410845957) from @srebhan, in my opinion the method is complex enough to keep it separate.

Reply:
That is a different thing, there the question was to include all these separate functions inside the parseconfig function, which I agree would be too complex.

I don't think it would become complex to have the content of parseConfig inside Init, who is currently just calling parseConfig. So I still think these 2 can be merged.. The functions like `parsePackageMetrics`, `parseCPUMetrics` and so on can stay as is for me.

Reply:
The main purpose to have it is that as for now all exported methods look short and clean. `Init` has now tow cleanly separated tasks: disable metrics if not supported by the CPU, and parse them along with the rest of config parameters. This separation between disabling metrics and parsing parameters allows to better unit test these scenarios. Right now we can test parsing logic isolated from the remaining logic. Merging these functions would decrease readability on the function and increase complexity in unit tests. Since this is not related to functionality, could you consider to  keep this methods as they are?